juju/framework.py

```diff
@@ -537,8 +548,15 @@ def __init__(self, parent, attr_name):
         except NoSnapshotError:
             data = StoredStateData(parent, attr_name)
 
+        # __dict__ is used to avoid infinite recursion.
         self.__dict__["_data"] = data
         self.__dict__["_attr_name"] = attr_name
+        self.__dict__["handle"] = handle
+
+        parent.framework.observe(parent.framework.on.pre_commit, self)
```

Comment:
This will create a new observer every time a BoundStoredState is created, and we're not doing a good job of cleaning this map right now.

Reply:
Right, as discussed we will need to use weak references for observers and clean them up if they are gone.